dex/candles/candles.go

```diff
@@ -101,10 +101,27 @@ func (c *Cache) WireCandles(count int) *msgjson.WireCandles {
 	return wc
 }
 
-// Delta calculates the the change in rate, as a percentage, and total volume
-// over the specified period going backwards from now. Because the first candle
-// does not necessarily align with the cutoff, the rate and volume contribution
-// from that candle is linearly interpreted between the endpoints. The caller is
+// CandlesCopy returns a deep copy of Candles with the oldest candle at the
+// first index.
+func (c *Cache) CandlesCopy() []msgjson.Candle {
+	sz := len(c.Candles)
+	candles := make([]msgjson.Candle, sz)
+	switch {
+	case sz == 0:
+	case sz == c.cap: // circular mode
+		oldIdx := c.cursor + 1
+		copy(candles, c.Candles[oldIdx:])
+		copy(candles[sz-oldIdx:], c.Candles)
```

Comment:
Is this wrong if it is called just after the size becomes c.cap but before it starts to reuse index 0?

Reply:
If len is 1000 then cursor is at 999... so ok I think.

Reply:
Looks to work right.  Could make some unit tests to explore the edges to be sure about it.  Also we could wrap like done in `Add` with `oldIdx := (c.cursor + 1) % c.cap` so `oldIdx` really is the index of the oldest candle as the name implies?